src/EditorFeatures/Core/Implementation/KeywordHighlighting/AbstractKeywordHighlighter.cs

```diff
@@ -12,67 +13,86 @@ internal abstract class AbstractKeywordHighlighter<TNode> : AbstractKeywordHighl
 
-        protected sealed override IEnumerable<TextSpan> GetHighlightsForNode(SyntaxNode node, CancellationToken cancellationToken)
-            => GetHighlights((TNode)node, cancellationToken);
+        protected sealed override void AddHighlightsForNode(SyntaxNode node, List<TextSpan> highlights, CancellationToken cancellationToken)
+            => AddHighlights((TNode)node, highlights, cancellationToken);
 
-        protected abstract IEnumerable<TextSpan> GetHighlights(TNode node, CancellationToken cancellationToken);
+        protected abstract void AddHighlights(TNode node, List<TextSpan> highlights, CancellationToken cancellationToken);
     }
 
     internal abstract class AbstractKeywordHighlighter : IHighlighter
     {
+        private static readonly ObjectPool<List<TextSpan>> s_textSpanListPool = new ObjectPool<List<TextSpan>>(() => new List<TextSpan>());
+        private static readonly ObjectPool<List<SyntaxToken>> s_tokenListPool = new ObjectPool<List<SyntaxToken>>(() => new List<SyntaxToken>());
+
         protected abstract bool IsHighlightableNode(SyntaxNode node);
 
-        public IEnumerable<TextSpan> GetHighlights(
-            SyntaxNode root, int position, CancellationToken cancellationToken)
+        public void AddHighlights(
+            SyntaxNode root, int position, List<TextSpan> highlights, CancellationToken cancellationToken)
         {
-            foreach (var token in GetTokens(root, position))
+            var _1 = s_textSpanListPool.GetPooledObject();
+            var _2 = s_tokenListPool.GetPooledObject();
+
+            var tempHighlights = _1.Object;
+            var touchingTokens = _2.Object;
+            AddTouchingTokens(root, position, touchingTokens);
+
+            foreach (var token in touchingTokens)
             {
                 for (var parent = token.Parent; parent != null; parent = parent.Parent)
                 {
                     if (IsHighlightableNode(parent))
                     {
-                        var highlights = GetHighlightsForNode(parent, cancellationToken);
+                        tempHighlights.Clear();
+                        AddHighlightsForNode(parent, tempHighlights, cancellationToken);
 
-                        // Only return them if any of them matched
-                        if (highlights.Any(span => span.IntersectsWith(position)))
+                        if (AnyIntersects(position, tempHighlights))
                         {
-                            // Return the non-empty spans
-                            return highlights.Where(s => !s.IsEmpty).Distinct();
+                            highlights.AddRange(tempHighlights);
+                            return;
                         }
                     }
                 }
             }
+        }
 
-            return SpecializedCollections.EmptyEnumerable<TextSpan>();
+        private static bool AnyIntersects(int position, List<TextSpan> highlights)
```

Comment:
`highlights.Any(h => h.IntersectsWith(position))`

Reply:
that allocates :)  which is what the PR is trying to prevent :)

Reply:
I don't think it does. Assuming it hasn't changed a great deal over the years: https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/AnyAll.cs#L48

Reply:
hmm... I would have thought creating an IEnumerator could be smarter if backed by an ICollection. Agree with the lambda allocation.

Reply:
that would be nice.  but it's not :D